File: original artifact/Ian_Foss_cs-360_project_three_inventory_app/cs360_project_three/app/src/main/java/com/example/cs_360_project_three/InventoryAdapter.java
package com.example.cs_360_project_three;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class InventoryAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    public interface OnItemClickListener {
        void onItemClick(InventoryItemEntity item);
    }

    private static final int VIEW_TYPE_HEADER = 0;
    private static final int VIEW_TYPE_ITEM = 1;
    InventoryItemRepository repo;
    private List<InventoryItemEntity> itemList;
    private final OnItemClickListener clickListener;

    private Context context;

    public InventoryAdapter(Context context, List<InventoryItemEntity> itemList, InventoryItemRepository repo, OnItemClickListener clickListener) {
        this.itemList = itemList;
        this.context = context;
        this.repo = repo;
        this.clickListener = clickListener;
    }

    @Override
    public int getItemViewType(int position) {
        return (position == 0 ) ? VIEW_TYPE_HEADER : VIEW_TYPE_ITEM;
    }


    // Inflates views for recyclerview header row and item rows
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        if (viewType == VIEW_TYPE_HEADER) {
            View view = inflater.inflate(R.layout.recyclerview_header_row, parent, false);
            return new HeaderViewHolder(view);
        } else {
            View view = inflater.inflate(R.layout.recyclerview_inventory_row, parent, false);
            return new InventoryViewHolder(view);
        }
    }


    // Binds InventoryItem data to the viewholders
    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {

         // Ensure the header row does not get the data
         if (getItemViewType(position) == VIEW_TYPE_ITEM) {
             int dataPosition = position - 1;
             InventoryItemEntity item = itemList.get(dataPosition);
             ((InventoryViewHolder) holder).bind(item, () -> {
                repo.delete(item);
             });

             holder.itemView.setOnClickListener(v -> clickListener.onItemClick(item));
        }
    }


    @Override
    public int getItemCount() {
        return itemList.size() + 1 ;
    }


    public void setItems(List<InventoryItemEntity> newItems) {
        this.itemList = new ArrayList<>(newItems);
        notifyDataSetChanged();
    }



}

File: enhancements/enhancement-three/Enhancement Three/app/src/main/java/com/example/CS499_Capstone/InventoryAdapter.kt
package com.example.CS499_Capstone

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import kotlin.collections.ArrayList
import kotlin.collections.MutableList

class InventoryAdapter(
    private val context: Context,
    private var itemList: List<InventoryItemEntity>,
    var repo: InventoryItemRepository,
    private val clickListener: OnItemClickListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {


    interface OnItemClickListener {
        fun onItemClick(item: InventoryItemEntity)
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) VIEW_TYPE_HEADER else VIEW_TYPE_ITEM
    }


    // Inflates views for recyclerview header row and item rows
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.getContext())

        if (viewType == VIEW_TYPE_HEADER) {
            val view = inflater.inflate(R.layout.recyclerview_header_row, parent, false)
            return HeaderViewHolder(view)
        } else {
            val view = inflater.inflate(R.layout.recyclerview_inventory_row, parent, false)
            return InventoryViewHolder(view)
        }
    }


    // Binds InventoryItem data to the viewholders
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        // Ensure the header row does not get the data

        if (getItemViewType(position) == VIEW_TYPE_ITEM) {
            val dataPosition = position - 1
            val item = itemList.get(dataPosition)
            (holder as InventoryViewHolder).bind(item, Runnable {
                repo.delete(item)
            })

            holder.itemView.setOnClickListener(View.OnClickListener { v: View ->
                clickListener.onItemClick(
                    item
                )
            })
        }
    }


    override fun getItemCount(): Int {
        return itemList.size + 1
    }


    fun setItems(newItems: List<InventoryItemEntity>) {
        itemList = newItems.toMutableList()
        notifyDataSetChanged()
    }


    companion object {
        private const val VIEW_TYPE_HEADER = 0
        private const val VIEW_TYPE_ITEM = 1
    }
}
